client/mm/exchange_adaptor.go

```diff
@@ -3466,6 +3598,167 @@ func (u *unifiedExchangeAdaptor) Book() (buys, sells []*core.MiniOrder, _ error)
 	return u.CEX.Book(u.baseID, u.quoteID)
 }
 
+func (u *unifiedExchangeAdaptor) latestCEXProblems() *CEXProblems {
+	u.cexProblemsMtx.RLock()
+	defer u.cexProblemsMtx.RUnlock()
+	if u.cexProblems == nil {
+		return nil
+	}
+	return u.cexProblems.copy()
+}
+
+func (u *unifiedExchangeAdaptor) latestEpoch() *EpochReport {
+	reportI := u.epochReport.Load()
+	if reportI == nil {
+		return nil
+	}
+	return reportI.(*EpochReport)
+}
+
+func (u *unifiedExchangeAdaptor) updateEpochReport(report *EpochReport) {
+	u.epochReport.Store(report)
+	u.clientCore.Broadcast(newEpochReportNote(u.host, u.baseID, u.quoteID, report))
+}
+
+// tradingLimitNotReached returns true if the user has not reached their trading
+// limit. If it has, it updates the epoch report with the problems.
+func (u *unifiedExchangeAdaptor) tradingLimitNotReached(epochNum uint64) bool {
+	var tradingLimitReached bool
+	var err error
+	defer func() {
+		if err == nil && !tradingLimitReached {
+			return
+		}
+
+		u.updateEpochReport(&EpochReport{
+			PreOrderProblems: &BotProblems{
+				UserLimitTooLow: tradingLimitReached,
+				UnknownError:    err,
+			},
+			EpochNum: epochNum,
+		})
+	}()
+
+	userParcels, parcelLimit, err := u.clientCore.TradingLimits(u.host)
+	if err != nil {
+		return false
+	}
+
+	tradingLimitReached = userParcels >= parcelLimit
+	return !tradingLimitReached
+}
+
+type cexProblemType uint16
+
+const (
+	cexTradeProblem cexProblemType = iota
+	cexDepositProblem
+	cexWithdrawProblem
+)
+
+func (u *unifiedExchangeAdaptor) updateCEXProblems(typ cexProblemType, assetID uint32, err error) {
```

Comment:
Could this func be simplified?

Reply:
I don't disagree. The locking scheme seems like overkill. We'll follow up.